### ISSUE TITLE
Balance monster difficulty

### ISSUE THREAD

**Issue by mciul** [i7/kerkerkruip#356]
I'm using my new fighttest command (https://github.com/i7/kerkerkruip/commit/66284b084f46db45e4325f91aaa2d6977424b0f2) to see how various monsters measure up to each other. I'll record the stats here and we can consider whether any changes should be made.

In 100 fights, Israfel was killed 47 times and Bodmall was killed 53 times.

In 100 fights, the giant tentacle was killed 100 times and the minotaur was killed 0 times.

In 100 fights, the hound was killed 94 times and the demon of rage was killed 6 times.
In 100 fights, the hound was killed 29 times and the chain golem was killed 71 times.
In 100 fights, the hound was killed 84 times and the jumping bomb was killed 16 times.
In 100 fights, the hound was killed 6 times and the angel of compassion was killed 94 times.
In 100 fights, the chain golem was killed 99 times and the demon of rage was killed 1 times.
In 100 fights, the chain golem was killed 86 times and the jumping bomb was killed 14 times.
In 100 fights, the chain golem was killed 16 times and the angel of compassion was killed 84 times.
In 100 fights, the rusted chain golem was killed 75 times and the Reaper was killed 25 times.
In 100 fights, the jumping bomb was killed 6 times and the demon of rage was killed 94 times.

Notes:
- I don't think the jumping bomb's free dodge is given to other creatures, so it may show up as overpowered.
- The angel of compassion's radiance was not reset for each fight, so it got weaker each time it killed its opponent. Likewise, the demon of rage's power increased.
- If somebody gets disarmed, I don't think the AI will try to get their weapon back, and reviving doesn't get it back either. That didn't stop the Reaper from kicking chain golem butt, though!
- errr... in addition to blindness, it looks like rust might need to be reset when reviving too.
- These tests happen in an arena, so no axe powers for the Minotaur. Not a problem against the tentacle, apparently!
- I still need to work on groupings so Fanatics of Aite, mindslug, and overmind can have a fair chance
- stat changes won't get reset either, I think, so the wisps will have an advantage over time too.

**Comment by mciul** [comment]
Okay, a fresh start after adding some changes - now most things should be reset at the start of every battle, and I've changed the reviving process so it's easy to add more rules when they're needed, like with sudden combat reset. So here are the new numbers:

In 100 fights, The hound was killed 38 times and the demon of rage was killed 62 times.

**Comment by mciul** [comment]
Fixed the bug. New stats, including stats copied from above:

In 100 fights, The hound was killed 38 times and the demon of rage was killed 62 times.
In 100 fights, The hound was killed 31 times and the chain golem was killed 69 times.
In 100 fights, The hound was killed 78 times and the jumping bomb was killed 22 times.

**Comment by VictorGijsbers** [comment]
This is a great idea. I don't think we need all the monsters to be perfectly balanced, but this will certainly allow us to get a little more consistency. I'm also very curious whether we'll see any non-transitivity: cases where A beats B, B beats C, and C beats A. I certainly hope so!

I don't think we need to test the wisps of pain. They'll automatically lose any fight since they cannot deal damage; the challenge with the wisps is to kill them without losing too much stat points.

**Comment by mciul** [comment]
If we don't find any non-transitivity, maybe we can come up with a monster who can provide it! :)

**Comment by curiousdannii** [comment]
This is an artificial situation, as what matters is whether they are balance in respect to the PC. Still, if we found that one was defeating another 95% of the time that would indicate that something probably needs to change.

I'd still love to eventually come up with a system where we can have a web server recording statistics from playthroughs. (Lots of interpreter hacking.) We could then see which rogues kill the most players and which get defeated too easily.

**Comment by mciul** [comment]
Yes, that would be best!

I've been considering writing an AI for the PC, but I'm not sure if that's really worth the effort.

**Comment by VictorGijsbers** [comment]
I don't think it would be worth the effort, since monsters are also balanced in some other ways, e.g., bu the reward you'll get from them. I have a feeling that Israfel might be the toughest level 4 monster. But that's not so bad -- he'll also give you the best level 4 powers. So you can afford to spend some more resources fighting him. It might actually be unbalanced to make Israfel weaker!

Of course, if the automatic fights show that Israfel is actually weaker than the other level 4 monsters, then we might need to make some changes.

**Comment by curiousdannii** [comment]
To really test against a PC AI, the PC would need other powers. It would need to test combinations of powers, which would just get far to complex. And ultimately we don't want a completely even game for every possible combination of powers, because using those powers well is part of the challenge. Some games you get lucky and some you get unlucky.

**Comment by mciul** [comment]
You sure do... I recently game where Malygris had unghoulification powers, pretty fast healing, and no way to stop him from teleporting around...

**Comment by mciul** [comment]
As for Israfel, I agree that it's probably the hardest to beat, but in my tests, it seems pretty evenly matched with the Fanatics of Aite, which I think are the easiest. On the other hand, the player has a lot of strategy options with the Fanatics that Israfel might not use.

I don't think I'll spend too much time testing the Overmind against the other Level 4 monsters - like the wisps of pain, the challenge isn't so much in killing him directly but in working with the limitations he imposes.

**Comment by mciul** [comment]
In 100 fights, Israfel was killed 49 times and the healer of Aite was killed 51 times.
In 100 fights, Bodmall was killed 16 times and the healer of Aite was killed 81 times.

You'll notice that there were three fights between Bodmall and the fanatics where no kill was recorded. I looked at the transcript and it appears that bees or thorns were the killers in three cases. Since it wasn't a hit, the killing rules didn't trigger. How should we handle that? In this specific case, it seems like the person creating the brambles should be the killer-guy, but what about other sources of damage? Can this be generalized? Should it?

**Comment by mciul** [comment]
Level 3 round robin:

In 100 fights, The mindslug was killed 0 times and the giant tentacle was killed 100 times.
In 100 fights, The mindslug was killed 41 times and the minotaur was killed 59 times.
In 100 fights, The giant tentacle was killed 92 times and the minotaur was killed 8 times.

This setup has the mindslug fighting with Fafhrd and Mouser. Contrary to what I thought originally, the minotaur IS able to maze people (mazing is banned in Arena of the Gods, but not generally). Which brings up an interesting possibility. We could use a modified fighttest to generate a table of battle results between the minotaur and other NPCs - and then use that table to calculate the result of the minotaur mazing every other character! There's also a health loss that goes with that battle... perhaps it's best to leave things as they are, although they're hacky.

**Comment by mciul** [comment]
Level 2 revisited, including some results copied from above:

In 100 fights, The chain golem was killed 75 times and the jumping bomb was killed 25 times.
In 100 fights, The chain golem was killed 73 times and the Reaper was killed 27 times.
In 100 fights, the chain golem was killed 99 times and the demon of rage was killed 1 times.
In 100 fights, The chain golem was killed 75 times and the hound was killed 25 times.
In 100 fights, the chain golem was killed 16 times and the angel of compassion was killed 84 times.
In 100 fights, The jumping bomb was killed 16 times and the Reaper was killed 84 times.
In 100 fights, The jumping bomb was killed 14 times and the demon of rage was killed 86 times.
In 100 fights, the hound was killed 94 times and the demon of rage was killed 6 times.
In 100 fights, the hound was killed 84 times and the jumping bomb was killed 16 times.
In 100 fights, the hound was killed 6 times and the angel of compassion was killed 94 times.

**Comment by VictorGijsbers** [comment]
How did you handle the demon and the angle? Did you reset their levels after each fight? If so, to what?

**Comment by mciul** [comment]
I created a "backups" table where I can save a monster's stats on demand. When a monster is revived, a rule checks the backups table and restores the stats it finds in there.

The code is here:

https://github.com/i7/kerkerkruip/blob/master/Kerkerkruip.materials/Extensions/Victor%20Gijsbers/Kerkerkruip%20Tests.i7x#L366

I've also made reviving into a rulebook, so any monster that has special properties can reset them when reviving.

As I've run the tests so far, the angel and the demon are backed up before anyone dies. Yesterday I was working on testing them after they've witnessed one death, but I ran into some problems I haven't figured out yet.

**Comment by mciul** [comment]
I've worked out a way to test the angel and the demon after they've witnessed a death. I added some commands for setting and viewing backups

> sneak
> battle wisps
> battle rage
> battle compassion
> smite wisps
> backups
> backup rage
> backup compassion
> backups
> smite rage
> smite compassion
> fighttest...

After this, I ran these fight tests again.

In 100 fights, The angel of compassion was killed 26 times and the chain golem was killed 74 times.
In 100 fights, The angel of compassion was killed 92 times and the jumping bomb was killed 8 times.
In 100 fights, The angel of compassion was killed 1 times and the Reaper was killed 99 times.
In 100 fights, The angel of compassion was killed 0 times and the demon of rage was killed 100 times
In 100 fights, The angel of compassion was killed 2 times and the hound was killed 98 times.

In 100 fights, The demon of rage was killed 18 times and the chain golem was killed 82 times.
In 100 fights, The demon of rage was killed 81 times and the jumping bomb was killed 19 times.
In 100 fights, The demon of rage was killed 41 times and the Reaper was killed 59 times.
In 100 fights, The demon of rage was killed 46 times and the hound was killed 54 times.

Looking at the demon of rage, I noticed I missed a lot of things that could potentially affect its performance. It really gets a lot of modifications! flying, heat protection, damage reduction, hit protection... I'm not sure it's worth it to try to control all of them. I'm going to think about what should happen when it gets revived, though - should demon-of-rage number be reset to 0? Should the other properties be reset?

**Comment by curiousdannii** [comment]
Rather than resetting them with code, would it be better to use the undo system? The in progress stats could either be stored in an external file, or with a protected memory block.

**Comment by mciul** [comment]
That's a good idea. I've already done most of that work with the automated testing system. At first I was concerned about collisions when using both systems together, but actually I don't think there's ever a need to do that.

**Comment by mciul** [comment]
Answer to self: Yes, I did reset protection when reviving: https://github.com/i7/kerkerkruip/commit/567c10721d71cbe41203e234c82dfe14fbb58785

**Comment by mciul** [comment]
TODO: create Table of Fight Statistics, to be saved in an external file

Columns:

Winner
Loser
Winner-variant: a number that serves as a unique id for every variant of a creature that we track. This includes demon of rage powerup level, angel of compassion radiance, and a unique id for each weapon that Miranda and the Reaper start with. Malygris has lots of variations, but I'm not sure how much we'll test him. Most creatures will use 0 or 1 all the time.
Loser-variant
Victory count
Final health - average
Final health - minimum
Final health - maximum

I'm guessing there's a formula that can be used to generate a series of die rolls to achieve the average, minimum, and maximum...? Any stats people out there who know?

**Comment by curiousdannii** [comment]
Btw, I was wanting to create a kind of setup rulebook, which would handle all the natural weapon rules and other when play begins rules. I was thinking we should have a phrase for giving rogues their weapons/items that will clone them if necessary, so that in the arena of the fallen it's not an issue if the player already has their weapon. (Or is that already the case?) Your reviving rules could all be put in the same rulebook.

**Comment by mciul** [comment]
I've been thinking about that too! I thought maybe I'd mentioned it somewhere, but I'm not sure. There is this: #321

**Comment by mciul** [comment]
Dannii, I looked back at my Automated Testing code and realized that I scrapped the undo saving code because I had too many problems with it. Instead I put a queue of tests in an external file and restart the game. That might work for fight tests too. But I think we need to implement monster kits first.

**Comment by VictorGijsbers** [comment]
The addict's amulet getting uncursed is certainly not intentional!

**Comment by VictorGijsbers** [comment]
Starting Kit in Kerkerkruip Monsters makes perfect sense, as does leving the player setup in S&F. We can add comments poiting to the other file.

Monster Abilities and Monsters were pried apart because Monsters is REALLY long. We could potentially take the general code out of Monsters and put it in Monster Abilities; and if necessary, we can at some point split Monsters. But I wouldn't merge them. (For me, at least, working with several smaller files is better than having one huge one.)

**Comment by mciul** [comment]
Here's an update on the tentacle after all of the changes I made to the damage system:

In 100 fights, The giant tentacle was killed 68 times and the minotaur was killed 32 times.

**Comment by VictorGijsbers** [comment]
What's the status of this? Anything we need to do before we can release?

**Comment by mciul** [comment]
I would suggest checking on the hound and the tentacle. Do others also feel that the hound is too difficult and the tentacle is too easy?

**Comment by rvdpluijm** [comment]
Hmm, I do find the tentacle too easy. Being a tentacle; could we perhaps add an ability that, when you do more than half the total amount of hitpoints in nonmagical damage, the damaged tentacle part becomes an extra tentacle with its own intelligence, more or less like the flatworm?

**Comment by mciul** [comment]
What do you think of the tentacle's new throwing ability, Remko?

I am improving fight tests right now as I test the Angel of Mercy. The Angel seems a bit underpowered at the moment - all of the other Level 3 monsters beat it more than half the time. But what really surprised me is that the tentacle basically always wins against it. I think that's because the tentacle is eyeless, so radiance makes no difference.

**Comment by mciul** [comment]
Remko, I have been thinking about adding a hydra-like monster that multiplies when you damage it. I thought maybe a big undead creature, to give the player a better chance of saving themselves when reading a scroll of summoning as a reaction. It could be like the rotting corpse, only the limbs that fall off keep attacking you!

**Comment by mciul** [comment]
My new fight tests are much faster. I'm going to post a few results here...

in 100 fights, giant tentacle won 94 times, angel of mercy won 6 times

**Comment by mciul** [comment]
I realized that the angel of mercy is pretty easy to beat if you simply unready your weapon. So I tweaked the AI - now the angel will take the gloves off sooner if it realizes there are no weapons to steal. I'm going to run the tests again, will post them in this comment...

**Comment by rvdpluijm** [comment]
I'll have to check it out. Will probably have time tonight to test the new build (I received my new MacBook Air and I still need to setup my environment (i.e. Github etc) or I could test this one on my Linux environment).

**Comment by mciul** [comment]
Current fight stats:

```
giant tentacle won 48 times, angel of mercy won 52 times
```

**Comment by curiousdannii** [comment]
Would it be useful to eventually run these on Travis?

**Comment by mciul** [comment]
I don't think so, although I've considered using them as source data for a more accurate minotaur-fight calculation

**Comment by mciul** [comment]
I'm running some Level 2 vs Level 3 comparisons now:

```
angel of compassion won 18 times, angel of mercy won 82 times
```

**Comment by mciul** [comment]
Level 3 vs Level 4:

Vs angel of mercy:
```
Israfel won 99 times, angel of mercy won 1 time
Bodmall won 28 times, angel of mercy won 72 times
healer of Aite won 36 times, angel of mercy won 64 times
Automatos won 93 times, angel of mercy won 7 times
```

Vs tentacle:
```
Israfel won 100 times, giant tentacle won 0 times
Bodmall won 100 times, giant tentacle won 0 times
healer of Aite won 100 times, giant tentacle won 0 times
Automatos won 97 times, giant tentacle won 3 times
```

Vs mindslug:
```
Israfel won 94 times, mindslug won 6 times
Bodmall won 100 times, mindslug won 0 times
healer of Aite won 100 times, mindslug won 0 times
Automatos won 88 times, mindslug won 12 times
```

Vs minotaur:
```
Israfel won 93 times, minotaur won 7 times
Bodmall won 85 times, minotaur won 15 times
healer of Aite won 93 times, minotaur won 7 times
Automatos won 75 times, minotaur won 25 times
```

**Comment by mciul** [comment]
Level 4 round robin:

```
Bodmall won 46 times, Israfel won 54 times
healer of Aite won 57 times, Israfel won 43 times
Automatos won 51 times, Israfel won 49 times
healer of Aite won 49 times, Bodmall won 51 times
Automatos won 43 times, Bodmall won 57 times
Automatos won 44 times, healer of Aite won 56 times
```

**Comment by mciul** [comment]
Level 2 vs level 1:

Vs demon of rage:
```
swarm of daggers won 54 times, demon of rage won 46 times
blood ape won 19 times, demon of rage won 81 times
Miranda won 59 times, demon of rage won 41 times
ravenous armadillo won 55 times, demon of rage won 45 times
```

vs angel of compassion:
```
swarm of daggers won 28 times, angel of compassion won 72 times
blood ape won 0 times, angel of compassion won 100 times
Miranda won 0 times, angel of compassion won 100 times
ravenous armadillo won 0 times, angel of compassion won 100 times
```

vs chain golem:
```
swarm of daggers won 62 times, chain golem won 38 times
blood ape won 44 times, chain golem won 56 times
Miranda won 36 times, chain golem won 64 times
ravenous armadillo won 19 times, chain golem won 81 times
```

vs hound:
```
swarm of daggers won 18 times, hound won 82 times
blood ape won 10 times, hound won 90 times
Miranda won 26 times, hound won 74 times
ravenous armadillo won 15 times, hound won 85 times
```

vs Reaper:
```
swarm of daggers won 29 times, Reaper won 66 times
blood ape won 16 times, Reaper won 84 times
Miranda won 27 times, Reaper won 73 times
ravenous armadillo won 16 times, Reaper won 84 times
```

**Comment by mciul** [comment]
Tweaked chain golem and swarm of daggers. Here are their new pairings. I'm counting the demon of rage as a Level 1 creature:

Level 1 vs chain golem:
```
swarm of daggers won 43 times, chain golem won 57 times
blood ape won 23 times, chain golem won 77 times
Miranda won 10 times, chain golem won 90 times
ravenous armadillo won 10 times, chain golem won 90 times
demon of rage won 30 times, chain golem won 70 times
```

Level 2 vs swarm of daggers
```
hound won 83 times, swarm of daggers won 17 times
Reaper won 71 times, swarm of daggers won 26 times
jumping bomb won 0 (96) times, swarm of daggers won 4 times
```

Level 2 round robin (excluding Angel of Compassion and demon of rage)
```
Reaper won 49 times, chain golem won 34 (51) times ("rusted chain golem" didn't count as the chain golem for my script - gotta fix that)
jumping bomb won 0 (89) times, chain golem won 11 times
jumping bomb won 0 (93) times, Reaper won 7 times
```

Level 1 round robin
```
blood ape won 31 times, swarm of daggers won 69 times
Miranda won 46 times, swarm of daggers won 54 times
ravenous armadillo won 86 times, swarm of daggers won 14 times
demon of rage won 56 times, swarm of daggers won 44 times
Miranda won 75 times, blood ape won 25 times
blood ape won 19 times, demon of rage won 81 times (copied from above)

Miranda won 59 times, demon of rage won 41 times (copied from above)

ravenous armadillo won 55 times, demon of rage won 45 times (copied from above)
```

@VictorGijsbers, do you remember why you boosted the jumping bomb last year? You gave it a +1 melee. It hardly seems like it needed it. In the games I've played, if I fight the jumping bomb with a Level 1 power, the bomb never misses (not counting the free dodge of course).

**Comment by mciul** [comment]
I improved the tests - I gave other monsters the bomb's free dodge when testing, and I discovered that the difficulty setting was arbitrarily high... so I think most of these results are useless. I'm going to delete most of the results from the comments above and start over...

**Comment by mciul** [comment]
Table of fight results:
|swarm of daggers|blood ape|ravenous armadillo|Miranda|chain golem|jumping bomb|Reaper|demon of rage|hound|angel of compassion|mindslug|giant tentacle|minotaur|angel of mercy|healer of Aite|Bodmall|overmind|Israfel|Automatos|Malygris
-------|----------------|---------|------------------|-------|-----------|------------|------|-------------|-----|-------------------|--------|--------------|--------|--------------|--------------|-------|--------|-------|---------|--------
swarm of daggers|X|64|15|54|50|51|34|55|14|23|0|5|1|0|0|0|10|0|0|0
blood ape|36|X|23|39|15|44|18|27|8|0|1|8|0|0|0|0|5|0|0|0
ravenous armadillo|85|77|X|75|21|8|14|57|23|0|1|100|0|0|0|0|23|0|0|0
Miranda|46|61|25|X|10|36|29|54|18|0|0|7|1|0|0|0|4|0|0|0
chain golem|50|85|79|90|X|78|56|82|43|29|0|10|6|27|0|0|49|6|1|0
jumping bomb|49|56|92|64|22|X|71|69|34|35|0|23|6|34|0|1|41|4|10|0
Reaper|66|82|86|71|44|29|X|69|34|0|0|20|6|0|0|0|33|0|2|0
demon of rage|45|73|43|46|18|31|31|X|21|0|1|3|0|1|0|0|2|0|0|0
hound|86|92|77|82|57|66|66|79|X|0|6|36|10|3|0|0|24|0|2|0
angel of compassion|77|100|100|100|71|65|100|100|100|X|75|68|54|12|1|34|97|8|17|0
mindslug|100|99|98|100|100|100|100|99|94|25|X|100|57|44|0|0|100|4|10|0
giant tentacle|95|92|0|93|90|77|80|97|64|32|0|X|32|50|0|0|21|0|0|0
minotaur|99|100|100|99|94|94|94|100|90|46|43|68|X|45|12|18|67|8|32|0
angel of mercy|100|100|100|100|73|66|100|99|97|88|56|50|55|X|68|67|99|0|5|0
healer of Aite|100|100|100|100|100|100|100|100|100|99|100|100|88|32|X|38|100|64|56|0
Bodmall|100|100|100|100|100|99|100|100|100|66|100|100|82|33|62|X|100|47|61|0
overmind|90|95|77|96|51|59|67|98|76|3|0|79|33|1|0|0|X|0|2|0
Israfel|100|100|100|100|94|96|100|100|100|92|96|100|92|100|36|53|100|X|50|0
Automatos|100|100|100|100|99|90|98|100|98|83|90|100|68|95|44|38|98|50|X|0
Malygris|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|X


Rankings:
monster|performance|weakest defeater|losses to defeater|strongest defeated|wins to defeated
-------|-----------|----------------|------------------|------------------|----------------
blood ape|11%|blood ape||mindslug|1
Miranda|15%|blood ape|39|minotaur|1
demon of rage|16%|blood ape|27|angel of mercy|1
swarm of daggers|19%|blood ape|36|minotaur|1
ravenous armadillo|25%|blood ape|23|mindslug|1
Reaper|28%|blood ape|18|Automatos|2
jumping bomb|32%|blood ape|44|Israfel|4
hound|36%|blood ape|8|Automatos|2
chain golem|36%|blood ape|15|Israfel|6
giant tentacle|43%|blood ape|8|angel of mercy|50
overmind|43%|blood ape|5|Automatos|2
angel of compassion|62%|swarm of daggers|23|Israfel|8
minotaur|63%|Miranda|1|Israfel|8
mindslug|64%|blood ape|1|Israfel|4
angel of mercy|69%|demon of rage|1|healer of Aite|68
Bodmall|81%|jumping bomb|1|Israfel|47
Automatos|81%|Reaper|2|Israfel|50
healer of Aite|83%|angel of compassion|1|Israfel|64
Israfel|84%|jumping bomb|4|Israfel|
Malygris|100%|Malygris||Malygris|

**Comment by mciul** [comment]
Check out the tables above! Because of the massive speedups in fight tests I was able to run a full round robin over the course of a couple days. I tallied the total victories for each monster and generated a table of ranking from least to most victories, along with the lowest-ranked monster to defeat each one and the hightest-ranked monster that each one defeated, along with how many times those things happened.

I used the map legend characters for each monster in the header of the first table. Automatos and the Angel of Mercy don't have legend characters... who has been making those? I came up with suggestions for each of them - "y" for Angel of Mercy and "U" for Automatos.

Based on this chart I would say that the hound might still be a little overpowered (and since I boosted the chain golem, it is too). The tentacle is bordering on powerful enough but could still use a small boost I think. And the angel of mercy might stand to be dialed back very slightly.

The armadillo fared surprisingly well. I'm not sure what to do about that. I find the armadillo both boring and frustrating to fight. Can anyone think of an entertaining handicap for it? Maybe it could start out weaker and get stronger every time it eats items.

The armadillo turns into a vampire when it eats Drakul's lifeblood. Can we have it do more crazy things like that? Essences should work on it, for example... I'm fine with waiting for Kerkerkruip 11 to do it.

**Comment by mciul** [comment]
I've tweaked several monsters including the armadillo, the minotaur, the tentacle, Miranda, the chain golem, and the hound. The tentacle is still scoring low as a level 3 monster but it's much stronger and has a new ability- sometimes hitting it will not break its grip, if the damage is low. The hound has been reduced in power and the chain golem has been boosted. They are now scoring about equal to each other and slightly better than the jumping bomb. The minotaur himself has not been changed but his axe has been boosted along with all other axes, according to #374.

I left the blood ape alone because it seemed to make sense to have it start out weak. Miranda has been boosted very slightly and the armadillo has been dialed back to bring them closer to the other Level 1 monsters. This was in line with my experience that fights with Miranda were too short and fights with the armadillo were a little long.

The whirling chains are now quite a bit harder to get past, but they do less damage when they hit you. The chain golem's health is lower. The hound's teeth have a lower damage die and the hound has slightly less health and body score.

Table of fight results:

|†|A|a|m|g|j|R|d|h|C|s|L|t|y|H|B|O|I|U|M
-------|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
swarm of daggers|X|65|27|53|50|46|36|37|17|35|0|0|0|2|0|0|13|3|0|0
blood ape|35|X|47|25|18|34|18|35|22|0|1|2|0|0|0|0|6|0|0|0
ravenous armadillo|73|53|X|60|15|8|9|51|22|0|1|79|0|0|0|0|12|0|0|0
Miranda|47|75|40|X|36|39|31|58|33|0|1|2|1|0|0|0|4|0|0|0
chain golem|50|82|85|64|X|80|40|70|44|19|0|0|2|38|0|0|42|5|1|0
jumping bomb|54|66|92|61|20|X|54|60|19|46|0|20|19|37|0|0|38|0|6|0
Reaper|64|82|91|69|60|46|X|72|37|0|3|8|4|1|0|0|22|0|2|0
demon of rage|63|65|49|42|30|40|28|X|30|0|1|0|4|0|0|0|7|0|0|0
hound|83|78|78|67|56|81|63|70|X|1|2|5|6|0|0|0|17|0|1|0
angel of compassion|65|100|100|100|81|54|100|100|99|X|71|25|82|21|1|26|97|15|26|0
mindslug|100|99|99|99|100|100|97|99|98|29|X|90|47|40|0|0|97|9|6|0
giant tentacle|100|98|21|98|100|80|92|100|95|75|10|X|47|98|0|0|68|3|0|0
minotaur|100|100|100|99|98|81|96|96|94|18|53|53|X|55|15|1|90|4|11|0
angel of mercy|98|100|100|100|62|63|99|100|100|79|60|2|45|X|49|51|94|3|2|0
healer of Aite|100|100|100|100|100|100|100|100|100|99|100|100|85|51|X|45|100|60|51|0
Bodmall|100|100|100|100|100|100|100|100|100|74|100|100|99|49|55|X|100|42|70|0
overmind|87|94|88|96|58|62|78|93|83|3|3|32|10|6|0|0|X|0|1|0
Israfel|97|100|100|100|95|100|100|100|100|85|91|96|96|97|40|58|100|X|43|0
Automatos|100|100|100|100|99|94|98|100|99|74|93|100|89|98|49|30|99|57|X|0
Malygris|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|X


Rankings:

monster|performance|weakest defeater|losses to defeater|strongest defeated|wins to defeated
-------|-----------|----------------|------------------|------------------|----------------
blood ape|12%|blood ape||mindslug|1
demon of rage|18%|blood ape|35|mindslug|1
Miranda|19%|blood ape|25|mindslug|1
ravenous armadillo|20%|blood ape|47|mindslug|1
swarm of daggers|20%|blood ape|35|Israfel|3
Reaper|29%|blood ape|18|Automatos|2
jumping bomb|31%|blood ape|34|Automatos|6
hound|32%|blood ape|22|Automatos|1
chain golem|32%|blood ape|18|Israfel|5
overmind|41%|blood ape|6|Automatos|1
giant tentacle|57%|blood ape|2|Israfel|3
angel of compassion|61%|swarm of daggers|35|Israfel|15
minotaur|61%|demon of rage|4|Israfel|4
angel of mercy|63%|swarm of daggers|2|Israfel|3
mindslug|63%|blood ape|1|Israfel|9
Automatos|83%|Reaper|2|Israfel|57
Bodmall|83%|angel of compassion|26|Israfel|42
healer of Aite|83%|angel of compassion|1|Israfel|60
Israfel|84%|swarm of daggers|3|Israfel|
Malygris|100%|Malygris||Malygris|

**Comment by mciul** [comment]
Victor, I forgot that discussion at the beginning about not making Israfel weaker, and I ended up slowing down its healing ability a little. It's still the toughest Level 4 monster but not as tough as it was. If anyone wants to do some playtesting and give feedback on any of the monsters at this point, that would be really helpful. Also, I haven't changed the powers to reflect changes in the monsters... is there a scheme for figuring out how that should work?

**Comment by curiousdannii** [comment]
Thank you for doing all of this!

**Comment by mciul** [comment]
It was my pleasure!

I played a game last night with the Angel of Mercy. It felt pretty easy, actually, but I'm considering leaving things the way they are for now: in some ways the Angel of Mercy is like the blood ape. If you let it steal your weapon, you can get a size-agnostic weapon with bigger damage after you kill it. So the player has a choice to make the battle harder for greater reward.

I've also decided to go ahead and apply the angel's special parrying effect to all monsters. (defender size does not affect attack strength when parrying) But I am going to add a weapon size penalty for parrying with a wrong-sized weapon.

**Comment by rvdpluijm** [comment]
I find the Angel of Mercy extremely easy, especially since, before her attack, she mentions that there is still time to flee. If one attacks immediately after that, there is a chance to break the concentration. I constantly got two turns to concentrate. Perhaps a little tweak in the AI is in order (admittedly, I played on the lowest level since I tried it during my break, but I cannot imagine we adapted the AI level).

**Comment by VictorGijsbers** [comment]
Awesome! Really, really good. I can't wait to start up the game and try the new balance!

We don't need to change the powers, I think -- they're not closely correlated to the monster stats. Does the angel of mercy already have a power?

**Comment by VictorGijsbers** [comment]
I just played two nice and satisfying games (one where I used the staff of insanity to make Malygris insane. Then one where I grabbed the rod of the master builder in the room where Malygtis was, at which point he teleported away and summoned the assassin. I had almost no powers... but managed to survive by using the holy symbols. Then I found a secret room, the Mausoleum, with two weak undead and a tomb that contained the Caduceus. Great weapon, with which I put everyone asleep!).

No bugs. Everything seemed balanced, good fights, about as difficult as I'd expect. But I haven't met the Angel of Mercy yet. :-)

**Comment by mciul** [comment]
That's great news!

Yes, the Angel of Mercy has a power. It's passive only, but defeating the Angel of Mercy gives you two great items.

I think I will boost the Angel of Mercy some more. My idea is for it to be a very strong monster that handicaps itself through a strict code of behavior. Currently I think the gauntlet of attraction is not powerful enough. I was so worried about it being overpowered that I think I went too far to limit it.

I'm currently trying to figure out how to handle the Angel's power during dreams, since it's a combat-only effect similar to the Blood Ape's, but it affects radiance, which is a bit more complicated to manage than size.

**Comment by mciul** [comment]
Remko, perhaps we should give the Angel of Mercy a concentration-breaking skill like Miranda and the Overmind. I think that would fit with its personality...

**Comment by mciul** [comment]
Here's the latest stats: I've made a lot of changes to the Angel of Mercy, and I've also made some AI changes: there is now a chance to win rule for radiance, everyone gets to skip the size bonus/penalty when parrying (there's a weapon size penalty instead), and the AI has been tweaked to adjust for that, although I didn't make any changes to the "best defence" checks.

Table of fight results:

 |†|A|a|m|g|j|R|d|h|C|s|L|t|y|H|B|O|I|U|M
-------|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---
swarm of daggers|†|62|16|52|55|56|41|47|20|34|0|0|0|1|0|0|11|2|0|0
blood ape|38|A|39|30|21|36|10|37|21|0|1|0|0|0|0|0|1|0|0|0
ravenous armadillo|84|61|a|62|14|11|10|48|20|0|0|80|0|0|0|0|19|0|0|0
Miranda|48|70|38|m|27|30|35|62|21|0|1|3|2|0|0|0|4|0|0|0
chain golem|45|79|86|73|g|74|37|66|43|14|0|2|3|23|0|0|36|7|1|0
jumping bomb|44|64|89|70|26|j|65|59|22|48|0|18|20|39|0|0|44|0|11|0
Reaper|59|90|90|65|63|35|R|67|34|1|2|6|7|0|0|0|19|0|3|0
demon of rage|53|63|52|38|34|41|33|d|31|0|0|2|1|1|0|0|0|0|0|0
hound|80|79|80|79|57|78|66|69|h|0|5|5|3|5|0|0|23|0|0|0
angel of compassion|66|100|100|100|86|52|99|100|100|C|55|31|98|20|4|49|99|7|20|0
mindslug|100|99|100|99|100|100|98|100|95|45|s|95|48|51|1|0|99|15|9|0
giant tentacle|100|100|20|97|98|82|94|98|95|69|5|L|56|86|0|0|77|5|1|0
minotaur|100|100|100|98|97|80|93|99|97|2|51|44|t|61|17|2|80|13|10|0
angel of mercy|99|100|100|100|77|61|100|99|95|80|49|14|39|y|15|51|97|4|5|0
healer of Aite|100|100|100|100|100|100|100|100|100|96|99|100|83|85|H|41|100|80|54|0
Bodmall|100|100|100|100|100|100|100|100|100|51|100|100|98|49|59|B|100|37|63|0
overmind|89|99|81|96|64|56|81|100|77|1|0|23|20|3|0|0|O|0|6|0
Israfel|98|100|100|100|93|100|100|100|100|93|84|95|86|96|20|63|100|I|33|0
Automatos|100|100|100|100|99|89|97|100|100|80|91|99|90|95|46|37|94|67|U|0
Malygris|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|100|M


Rankings:

monster|performance|weakest defeater|losses to defeater|strongest defeated|wins to defeated
-------|-----------|----------------|------------------|------------------|----------------
blood ape|12%|blood ape||mindslug|1
Miranda|17%|blood ape|30|mindslug|1
demon of rage|18%|blood ape|37|angel of mercy|1
swarm of daggers|20%|blood ape|38|Israfel|2
ravenous armadillo|21%|blood ape|39|giant tentacle|80
Reaper|28%|blood ape|10|Automatos|3
chain golem|31%|blood ape|21|Automatos|1
jumping bomb|32%|blood ape|36|Automatos|11
hound|33%|blood ape|21|mindslug|5
overmind|41%|blood ape|1|Automatos|6
giant tentacle|57%|Miranda|3|Automatos|1
minotaur|60%|Miranda|2|healer of Aite|17
angel of mercy|62%|demon of rage|1|healer of Aite|15
angel of compassion|62%|swarm of daggers|34|healer of Aite|4
mindslug|66%|blood ape|1|healer of Aite|1
Bodmall|81%|minotaur|2|healer of Aite|59
Israfel|82%|swarm of daggers|2|healer of Aite|20
Automatos|83%|Reaper|3|healer of Aite|46
healer of Aite|86%|minotaur|17|healer of Aite|
Malygris|100%|Malygris||Malygris|

**Comment by mciul** [comment]
I reduced the mindslug and the Fanatics, and brought Israfel right back where it was before I started tweaking. The Fanatics are still doing best out of the Level 4 monsters, and Automatos is very close with both, but let's play with it this way and see how it goes.

I've also adjusted the fight test scripts so I can build the table from multiple results files. The tables below are the combined results of 300 fights between every pairing of monsters, and hopefully they are more accurate as a result:

Table of fight results:

 |†|A|a|m|g|j|R|d|h|C|s|L|t|y|H|B|O|I|U|M
-------|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---
swarm of daggers|†|184|70|163|149|150|96|154|65|93|0|1|2|1|0|0|31|0|0|0
blood ape|116|A|116|93|78|107|37|105|59|0|4|0|0|1|0|0|8|0|0|0
ravenous armadillo|230|184|a|180|47|26|19|151|75|0|2|244|0|0|0|0|45|0|0|0
Miranda|137|207|120|m|68|100|97|187|81|0|7|7|4|1|0|0|20|0|0|0
chain golem|151|222|253|232|g|221|121|214|140|61|0|4|7|41|0|0|91|15|9|0
jumping bomb|150|193|274|200|79|j|180|188|73|129|0|55|65|156|0|0|116|5|24|0
Reaper|204|263|281|203|179|120|R|208|118|1|14|13|22|4|0|0|62|0|3|0
demon of rage|146|195|149|113|86|112|92|d|78|0|6|2|3|1|0|0|3|0|0|0
hound|235|241|225|219|160|227|182|222|h|0|20|13|14|5|0|0|62|0|1|0
angel of compassion|207|300|300|300|239|171|299|300|300|C|177|86|292|64|9|173|298|26|69|0
mindslug|300|295|298|293|300|300|285|294|280|121|s|258|122|144|0|1|296|32|21|0
giant tentacle|299|300|56|293|296|245|287|298|287|214|40|L|162|277|0|0|214|3|3|0
minotaur|298|300|300|296|293|235|278|297|286|8|177|138|t|182|62|3|260|26|26|0
angel of mercy|299|299|300|299|259|144|296|299|295|236|156|23|118|y|86|156|299|8|18|0
healer of Aite|300|300|300|300|300|300|300|300|300|291|300|300|238|214|H|97|300|232|112|0
Bodmall|300|300|300|300|300|300|300|300|300|127|299|300|297|143|203|B|300|113|194|0
overmind|269|292|255|280|209|184|238|297|238|2|3|86|40|1|0|0|O|0|9|0
Israfel|300|300|300|300|285|295|300|300|300|274|268|297|271|292|68|187|300|I|136|0
Automatos|300|300|300|300|291|276|297|300|299|231|278|297|274|282|188|106|291|164|U|0
Malygris|300|300|300|300|300|300|300|300|300|300|300|300|300|300|300|300|300|300|300|M


Rankings:

monster|performance|weakest defeater|losses to defeater|strongest defeated|wins to defeated
-------|-----------|----------------|------------------|------------------|----------------
blood ape|12%|blood ape||mindslug|4
demon of rage|17%|blood ape|105|mindslug|6
Miranda|18%|blood ape|93|mindslug|7
swarm of daggers|20%|blood ape|116|angel of compassion|93
ravenous armadillo|21%|blood ape|116|mindslug|2
Reaper|29%|blood ape|37|Automatos|3
chain golem|31%|blood ape|78|Automatos|9
hound|32%|blood ape|59|Automatos|1
jumping bomb|33%|blood ape|107|Automatos|24
overmind|42%|blood ape|8|Automatos|9
giant tentacle|57%|demon of rage|2|Automatos|3
minotaur|60%|demon of rage|3|healer of Aite|62
angel of mercy|62%|blood ape|1|healer of Aite|86
angel of compassion|63%|swarm of daggers|93|healer of Aite|9
mindslug|63%|blood ape|4|Automatos|21
Bodmall|82%|minotaur|3|healer of Aite|203
Israfel|83%|chain golem|15|healer of Aite|68
Automatos|83%|Reaper|3|healer of Aite|188
healer of Aite|83%|minotaur|62|healer of Aite|
Malygris|100%|Malygris||Malygris|

**Comment by mciul** [comment]
I just had an awesome game with the following monsters:

armadillo
wisps of pain
Angel of Compassion
hound
angel of mercy
Fanatics of Aite

I killed the armadillo, and fought the hound in the Temple of Nomos, sacrificing after I used the scales ability. I went to the quartering room and tried random scrolls until I summoned a malignant chanter and got ghoulified. This was a great combination! I killed the Angel of Mercy, the angel of compassion and the wisps, sacrificed the power of compassion, and fought the fanatics.

It's been a while since I reworked the damage system, but the Tormentor of Aite can't hurt you at all if you're undead! (I wonder if we should add some AI so she unreadies the staff of pain if her enemy is undead) I managed to kill the healer and the defender with the malignant chanter's help (he died in the process), then sacrificed the power of Mercy before killing the tormentor.

I went to fight Malygris. I didn't have any teleportation controllers, but I had the rod of the master builder, so I narrowed the path to the quartering room. I softened him up with grenades from the Laboratory then went downstairs to fight him. As soon as he teleported away, I headed over there so I could have the rotting corpse's help when the assassin showed up. I took a couple of hits but survived, then went to find Malygris again. Here's my favorite part: I used the gauntlet of attraction to steal the dagger of draining!

Malygris didn't survive long enough after that to really get to use it, but it was so satisfying.

The game may be slightly easy now, but I might also just be playing smarter. How does everyone else feel about it?

**Comment by VictorGijsbers** [comment]
That sounds like an awesome game. :-) I love the table of fighting scores. It's looking pretty good too, with all level 4 creatures around 82% (except the overmind, but that's fine); all level 3 creatures around 60%, all level 2 creatures around 30% and all level 1 creatures around 20%. One exception: the blood ape. I think I'll tweak it a bit so that his combat power increases a little more as he grows in size.
